### PR TITLE
Setup phased transition away from PyTorch version-specific handling of cuda availability and device counting

### PR DESCRIPTION
**Pull request by speediedan**
## What does this PR do?

Setup a phased transition away from PyTorch version-specific handling of CUDA availability (`torch.cuda.is_available()`) and device counting (`torch.cuda.device_count()`)

This is a follow-up to #15110 and [#85951](https://github.com/pytorch/pytorch/pull/85951) that prepares for removal of unnecessary PyTorch version-specific code once PyTorch 1.13 and 1.14 respectively are the minimum supported PyTorch.  

Specifically:
1. All local NVML-based CUDA device counting code copied from upstream PyTorch can be removed once PyTorch 1.13 is the minimum required.
2. The [_patch_cuda_is_available()](https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/lightning_lite/accelerators/cuda.py#L82) context manager (and associated usage of it) and redirection of [is_cuda_available](https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/lightning_lite/accelerators/cuda.py#L115) to [num_cuda_devices()](https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/lightning_lite/accelerators/cuda.py#L99) can be removed once PyTorch 1.14 is the minimum required.

As of PyTorch 1.14, [is_cuda_available()](https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/lightning_lite/accelerators/cuda.py#L115) and [num_cuda_devices()](https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/lightning_lite/accelerators/cuda.py#L99) should be functionally aliases for `torch.cuda.is_available()` and `torch.cuda.num_cuda_devices()` respectively that do not require special PyTorch-specific handling and can take advantage of potential upstream enhancements to the functions (e.g., if those functions abstract ROCm CUDA availability checks using ROCm's SMI analog to NVML).

It likely makes sense to retain the [is_cuda_available()](https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/lightning_lite/accelerators/cuda.py#L115) and [num_cuda_devices()](https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/lightning_lite/accelerators/cuda.py#L99) wrappers in the code base to facilitate future PyTorch version-specific handling if it becomes necessary but theoretically those wrappers could be replaced with their associated PyTorch functions as well.

Also perhaps relevant, I've tested these changes locally on PyTorch `1.13.0rc3` and `1.14.0.dev20221013`

cc @awaelchli @carmocca 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [X] Did you list all the **breaking changes** introduced by this pull request?
- [X] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

I made sure I had fun 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@speediedan Not sure if I 100% understand the proposed change. Is it because partial changes are in 1.13, where device_count uses NVML yet is_available is not yet using the new implementation?

**Comment by speediedan** [comment]
> @speediedan Not sure if I 100% understand the proposed change. Is it because partial changes are in 1.13, where device_count uses NVML yet is_available is not yet using the new implementation?

Yep! It allows to use the relevant PyTorch functions directly whenever they are available, only using the temporary PL versions when necessary.  It also allows us to remove the PL NVML code from upstream once PT 1.13 is the minimum (instead of when PT 1.14 is the minimum).

File: src/lightning_lite/plugins/precision/native_amp.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def __init__(
         if precision == "bf16" and not _TORCH_GREATER_EQUAL_1_10:
             raise ImportError("To use bfloat16 with native amp you must install torch greater or equal to 1.10.")
         if scaler is None and precision == 16:
-            with _patch_cuda_is_available():
+            with _patch_cuda_is_available():  # TODO: Remove once minimum supported PyTorch version is 1.14
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/native_amp.py:51]: The implementation already has this comment. No need to add it again here

```suggestion
            with _patch_cuda_is_available():
```

File: src/pytorch_lightning/plugins/precision/native_amp.py [hunk 2]
```diff
@@ -51,7 +51,7 @@ def __init__(
                 "To use bfloat16 with native amp you must install torch greater or equal to 1.10."
             )
         if scaler is None and precision == 16:
-            with _patch_cuda_is_available():
+            with _patch_cuda_is_available():  # TODO: Remove once minimum supported PyTorch version is 1.14
```
Comment (carmocca) [inline src/pytorch_lightning/plugins/precision/native_amp.py:54]: Same

File: src/pytorch_lightning/strategies/colossalai.py [hunk 3]
```diff
@@ -38,6 +38,7 @@
 
 _COLOSSALAI_AVAILABLE = RequirementCache("colossalai")
 if TYPE_CHECKING and _COLOSSALAI_AVAILABLE:
+    # TODO: Remove `_patch_cuda_is_available` usage for this strategy once minimum supported PyTorch version is 1.14
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/colossalai.py:41]: Same

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/plugins/precision/native_amp.py (hunk 2 above)
Comment (speediedan) [reply src/pytorch_lightning/plugins/precision/native_amp.py]: done, thanks!

File: src/pytorch_lightning/strategies/colossalai.py (hunk 3 above)
Comment (speediedan) [reply src/pytorch_lightning/strategies/colossalai.py]: done, thanks!

**Pull request merged**